.github/workflows/tag-publish.yaml

```diff
@@ -1,10 +1,170 @@
+---
 name: Publish release
-
 on:
-  workflow_dispatch:
+  workflow_dispatch: 
+    inputs:
+      artifact-tag:
+        description: "The tag associated with the artifact to deploy (eg. v1.2.3)."
+        type: string
+        required: true
+      environment:
+        description: "The publishing environment that the workflow should run in."
+        type: environment
+        required: true
+      # This is a workaround so that the actor who initiated a workflow run via a workflow dispatch event can determine the run ID of the started workflow run
+      workflow-tag:
+        description: "This field adds the provided value to a run step, allowing the calling actor to associate the started run with the GHA run ID."
+        type: string
+        required: false
+  release:
+    types:
+      - released  # This may need to be changed to "published"
+
+concurrency: "Limit to one build at a time for artifact tag ${{ inputs.artifact-tag || github.event.release.tag_name }}"
 
 jobs:
-  placeholder:
+  publish:
     runs-on: ubuntu-latest
+    # Unfortuntely this cannot be set as a global env var as they are not evaluated until runtime
+    environment: ${{ inputs.environment || 'release-prod' }}
+    permissions:
+      id-token: write
+    env:
+      ARTIFACT_TAG: ${{ inputs.artifact-tag || github.event.release.tag_name }}
+      LOCAL_ARTIFACTS_PATH: /tmp/artifacts
+      LOCAL_HELM_REPO_PATH: /tmp/helm-repo
+      LOCAL_TERRAFORM_REGISTRY_PATH: /tmp/terraform-registry
+      ENVIRONMENT_NAME: ${{ inputs.environment || 'release-prod' }}
     steps:
-      - run: echo "Placeholder"
+      - name: Validate artifact tag
+        env:
+          SEMVER_REGEX: ^v?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+        run: |
+          (echo "$ARTIFACT_TAG" | grep -qP "$SEMVER_REGEX") || (echo "The artifact tag $ARTIFACT_TAG is not a valid semver-coerced value"; exit 1)
+      - name: Validate environment
+        run: |
+          (echo "$ENVIRONMENT_NAME" | grep -qP '^publish-(prod|stage)$') || (echo "This workflow may only be ran from publishing environments"; exit 1)
```

Comment:
> But I feel this policy should not be encoded here

Where should it be implemented then?

IMO it's much better to be explicit here. Relying on the workflow failing due to a side effect of a variable not being available may result in part of the workflow running if somebody adds a variable with the same name to both the build and publishing environments.

> and this precludes developing a new environment and testing with that without altering the workflow

I see this as a good thing. A new environment would need a test run of affected workflows anyway, and this helps ensure that a test run is performed before the environment is allowed to be (normally) used.

> if the environment is spelt incorrectly

This _should_ be checked by GHA automatically due to the input `type` field value.

Reply:
> Where should it be implemented then?

My assumption of where this went is the cloud-terraform config, but I now realise this is controlling which roles can be assumed based on a repo/tag/branch/environment/etc.

I still feel that he bottom-level workflows are easier to read when they just do what they do and leave the "gatekeeping" to a higher level. I find all the auto-selecting environment setup code not easy to reason about with all the variations and branches, and trying to just read the core of what the workflow does - build plugins and upload them to a bucket - is hard to get a clear picture of.

I would be in favour of having a simple workflow that triggers only off a workflow_call event and is told what environment to use and what version to build. A higher level workflow would trigger off the various events and implement the github-specific logic that maps the event parameters to what the bottom-level workflow needs and passes it in. This is the approach we've been using in teleport.e - a "build" workflow that is called, and a "release" workflow which is dispatched. The top-level tag build workflow will just call all the "build" workflows when we get rid of drone and then we can think of getting rid of the "release" workflows (but we'll probably keep them because they are useful for testing to call manually).

Finally, if we do want to limit workflows to particular environments, let's think about a shared action that validates the environment:

    - name: Validate environment
      uses: shared-actions/validate-environment
      with:
        allowed-environments: publish-stage publish-prod
        environment: ${{ inputs.environment }}

> This should be checked by GHA automatically due to the input type field value

Hmmm. I don't use anything but the basic types supported by `workflow_call` (https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype). Sadly there is still not parity between the input parameters between dispatch and call. As I've been using that workflow split described above, workflow_call input parameter types become the limiting factor.

Reply:
> bottom-level workflows

Just to be clear while this technically is a bottom level workflow, it's also a top level workflow. That is to say, it is the only publishing workflow.

> I find all the auto-selecting environment setup code

This doesn't attempt to auto-select an environment - it is only used to _validate_ which environment has been selected by the caller (usually a release manager).

> I would be in favour of having a simple workflow that triggers only off a workflow_call event and is told what environment to use and what version to build

I generally try to avoid the "triggering workflow" -> "actionable workflow via call" pattern where it isn't already established because it adds a lot of boilerplate and sometimes makes it harder to follow where certain logic is defined.

That said, there are enough different triggers for the build workflow that it might make sense to break it up into the structure you're suggesting. This would eliminate, or at least simplify, most of the shell script in the build workflow `Set environment output values` step. What do you think?

> Finally, if we do want to limit workflows to particular environments, let's think about a shared action that validates the environment

Yea I'm all for that, especially now that we're beginning to standardize our environments/their purpose across the repos in our org. I've thought about this a bit and I think it might make sense to leave this step as is for now. Then we can try to schedule work next quarter (after we're entirely off Drone) to migrate this, and all the other similar logic that we've had this conversation about, to the shared-workflows repo next quarter. This would help ensure that this PR gets merged prior to the March 1 "off of Drone" deadline. If we want to go this route then I can file an issue somewhere to start tracking all these similar actions.

Reply:
Sounds good to me

Reply:
Let's get this in as is, and discuss in person strategies we want to use across the board and then get them all in line.